File: N-Queens_II.cpp
// Follow up for N-Queens problem.
// Now, instead outputting board configurations, 
// return the total number of distinct solutions.

class Solution {
public:
    bool check(vector<int> &queenPos, int row, int col, int n) {
        for (int i = 0; i < row; i++) {
            if (queenPos[i] == col) 
                return false; 
            if (abs(i-row) == abs(queenPos[i]-col)) 
                return false;
        }
        return true;
    }

    void putQueen(vector<int> &queenPos, int row, int n, int &result) {
        if (row == n) {
            result++;
            return;
        }
        for (int col = 0; col < n; col++) {
            if (check(queenPos, row, col, n)) {
                queenPos[row] = col;
                putQueen(queenPos, row+1, n, result);
                queenPos[row] = -1;
            }
        }
    }

    int totalNQueens(int n) {
        int result = 0;
        vector<int> queenPos(n, -1);
        putQueen(queenPos, 0, n, result);
        return result;
    }
};

File: N-Queens-II.py
class Solution:
    # @return an integer
    def totalNQueens(self, n):
        if n <= 0: return 0
        if n == 1: return 1
        queenPos = [-1 for i in range(n)]
        self.res = 0
        self.putQueen(queenPos, 0, n)
        return self.res
    
    def putQueen(self, queenPos, row, n):
        if row == n:
            self.res += 1
            return
        for col in range(n):
            if self.check(queenPos, row, col, n):
                queenPos[row] = col
                self.putQueen(queenPos, row+1, n)
                queenPos[row] = -1
                
    def check(self, queenPos, row, col, n):
        if col in queenPos: return False
        for i in range(row):
            if row+col == i+queenPos[i]: return False
            if row-col == i-queenPos[i]: return False
        return True
